File: code/zato-web-admin/src/zato/admin/web/views/definition/jms_wmq.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2019, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

# stdlib
import logging

# Zato
from zato.admin.web.forms import ChangePasswordForm
from zato.admin.web.forms.definition.jms_wmq import CreateForm, EditForm
from zato.admin.web.views import change_password as _change_password, CreateEdit, Delete as _Delete, Index as _Index, \
     method_allowed, ping_connection
from zato.common.odb.model import ConnDefWMQ

logger = logging.getLogger(__name__)

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'definition-wmq'
    template = 'zato/definition/jms-wmq.html'
    service_name = 'zato.definition.jms-wmq.get-list'
    output_class = ConnDefWMQ
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = ('cluster_id',)
        output_required = ('id', 'name', 'host', 'port', 'queue_manager', 'channel', 'cache_open_send_queues',
            'cache_open_receive_queues', 'use_shared_connections', 'ssl', 'ssl_cipher_spec', 'ssl_key_repository',
            'needs_mcd', 'max_chars_printed', 'username', 'use_jms')
        output_repeated = True

    def handle(self):
        return {
            'change_password_form': ChangePasswordForm(),
            'create_form': CreateForm(),
            'edit_form': EditForm(prefix='edit'),
        }

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = ('name', 'host', 'port', 'queue_manager', 'channel', 'cache_open_send_queues',
            'cache_open_receive_queues', 'use_shared_connections', 'ssl', 'ssl_cipher_spec', 'ssl_key_repository', 'needs_mcd',
            'max_chars_printed', 'username', 'use_jms')
        output_required = ('id',)

    def success_message(self, item):
        return 'Successfully {} IBM MQ definition `{}`'.format(self.verb, item.name)

class Create(_CreateEdit):
    url_name = 'definition-wmq-create'
    service_name = 'zato.definition.jms-wmq.create'

class Edit(_CreateEdit):
    url_name = 'definition-wmq-edit'
    form_prefix = 'edit-'
    service_name = 'zato.definition.jms-wmq.edit'

class Delete(_Delete):
    url_name = 'definition-wmq-delete'
    error_message = 'Could not delete IBM MQ definition'
    service_name = 'zato.definition.jms-wmq.delete'

@method_allowed('POST')
def ping(req, id, cluster_id):
    return ping_connection(req, 'zato.definition.jms-wmq.ping', id, 'IBM MQ definition')

@method_allowed('POST')
def change_password(req):
    return _change_password(req, 'zato.definition.jms-wmq.change-password')

File: code/zato-web-admin/src/zato/admin/static/js/definition/jms-wmq.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.ConnDefWMQ = new Class({
    toString: function() {
        var s = '<ConnDefWMQ id:{0} name:{1} cache_open_send_queues:{2}>';
        return String.format(s, this.id ? this.id : '(none)',
            this.name ? this.name : '(none)',
            this.cache_open_send_queues ? this.cache_open_send_queues : '(none)');
    }
});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.ConnDefWMQ;
    $.fn.zato.data_table.new_row_func = $.fn.zato.definition.jms_wmq.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'host', 'port', 'channel', 'max_chars_printed']);
})

$.fn.zato.definition.jms_wmq.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new IBM MQ definition', null);
}

$.fn.zato.definition.jms_wmq.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update IBM MQ definition', id);
}

$.fn.zato.definition.jms_wmq.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    var ssl = item.ssl == true;
    var ssl_cipher_spec = item.ssl_cipher_spec ? item.ssl_cipher_spec : '';
    var ssl_key_repository = item.ssl_key_repository ? item.ssl_key_repository : '';
    var needs_mcd = item.needs_mcd == true;
    var use_jms = item.use_jms == true;
    var username = item.username ? item.username : '';
    var queue_manager = item.queue_manager ? item.queue_manager : '';

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', item.host);
    row += String.format('<td>{0}</td>', item.port);

    row += String.format('<td>{0}</td>', item.queue_manager ? item.queue_manager : $.fn.zato.empty_value);
    row += String.format('<td>{0}</td>', item.channel);
    row += String.format('<td>{0}</td>', item.username ? item.username : $.fn.zato.empty_value);
    row += String.format('<td>{0}</td>', ssl ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.ssl_cipher_spec ? item.ssl_cipher_spec : $.fn.zato.empty_value);

    row += String.format('<td>{0}</td>', item.ssl_key_repository ? item.ssl_key_repository : $.fn.zato.empty_value);
    row += String.format('<td>{0}</td>', item.max_chars_printed);
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.data_table.change_password({0})'>Change password</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.definition.jms_wmq.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.definition.jms_wmq.delete_({0});'>Delete</a>", item.id));

    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.data_table.ping({0});'>Ping</a>", item.id));
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", item.cache_open_send_queues);
    row += String.format("<td class='ignore'>{0}</td>", item.cache_open_receive_queues);
    row += String.format("<td class='ignore'>{0}</td>", item.use_shared_connections);

    row += String.format("<td class='ignore'>{0}</td>", ssl);
    row += String.format("<td class='ignore'>{0}</td>", username);
    row += String.format("<td class='ignore'>{0}</td>", queue_manager);
    row += String.format("<td class='ignore'>{0}</td>", ssl_cipher_spec);
    row += String.format("<td class='ignore'>{0}</td>", ssl_key_repository);
    row += String.format("<td class='ignore'>{0}</td>", use_jms);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.definition.jms_wmq.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'IBM MQ definition `{0}` deleted',
        'Are you sure you want to delete IBM MQ definition `{0}`?',
        true);
}
